txpool/txpool.go

```diff
@@ -823,466 +463,712 @@ func (t *TxPool) validateTx(
 	return nil
 }
 
-// Underpriced checks whether given tx's price is less than any in remote transactions
-func (t *TxPool) Underpriced(tx *types.Transaction) bool {
-	lowestTx := t.remoteTxns.Pop()
-	if lowestTx == nil {
-		return false
+// addTx is the main entry point to the pool
+// for any newly received transaction. If the call to
+// addTx is successful an account is created
+// for this address (only once) and an addRequest is signaled.
+func (p *TxPool) addTx(origin txOrigin, tx *types.Transaction) error {
+	p.logger.Debug("add tx",
+		"origin", origin.String(),
+		"hash", tx.Hash.String(),
+	)
+
+	// validate incoming tx
+	if err := p.validateTx(tx); err != nil {
+		return err
 	}
-	// tx.GasPrice < lowestTx.Price
-	underpriced := tx.GasPrice.Cmp(lowestTx.price) < 0
-	if pushErr := t.remoteTxns.Push(lowestTx.tx); pushErr != nil {
-		t.logger.Error(fmt.Sprintf("Unable to push transaction to remoteTxn queue, %v", pushErr))
+
+	// check for overflow
+	if p.gauge.read()+slotsRequired(tx) > p.gauge.max {
+		return ErrTxPoolOverflow
 	}
 
-	return underpriced
-}
+	tx.ComputeHash()
 
-func (t *TxPool) Discard(slotsToRemove uint64, force bool) ([]*types.Transaction, bool) {
-	dropped := make([]*types.Transaction, 0)
-	for slotsToRemove > 0 {
-		if t.remoteTxns.Length() == 0 {
-			break
+	// check if already known
+	if _, ok := p.index.load(tx.Hash); ok {
+		if origin == gossip {
+			// silently drop known tx that is gossiped back
+			p.logger.Debug("dropping known gossiped transaction", "hash", tx.Hash.String())
+			return nil
+		} else {
+			return ErrAlreadyKnown
 		}
+	}
 
-		pricedTx := t.remoteTxns.Pop()
-		dropped = append(dropped, pricedTx.tx)
+	// initialize account queue for this address once
+	p.createAccountOnce(tx.From)
 
-		txSlots := slotsRequired(pricedTx.tx)
-		if slotsToRemove < txSlots {
-			return dropped, true
-		}
+	// send request [BLOCKING]
+	p.addReqCh <- addRequest{tx: tx, demoted: false}
 
-		slotsToRemove -= txSlots
+	return nil
+}
+
+// handleGossipTx handles receiving transactions
+// gossiped by the network.
+func (p *TxPool) handleGossipTx(obj interface{}) {
```

Comment:
Current behaviour is no different than adding a local transaction.

Reply:
I disagree with you on this one, at least since you have a check with the sealing flag at the beginning. So I'm still standing with my opinion that tests should be added here.

Reply:
[Added](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/a8651abb29dc10ffd471129264898bb40a4811cc)